### PR TITLE
Move alloc copy to finalize

### PR DESCRIPTION
**Pull request by scxiao**

### CODE REVIEW

File: src/targets/gpu/include/migraphx/gpu/hip.hpp [hunk 1]
```diff
@@ -151,7 +152,9 @@ struct hip_copy
     std::ptrdiff_t output_alias(const std::vector<shape>&) const { return 1; }
 };
 
-struct hip_load_memory
+void store_preallocated_param(context& ctx, const std::string& id, const argument& a);
```
Comment (pfultz2) [inline src/targets/gpu/include/migraphx/gpu/hip.hpp:155]: Whats the reason to have this in a seperate function? Is the `context` an incomplete type?

Comment (scxiao) [reply src/targets/gpu/include/migraphx/gpu/hip.hpp:155]: Yes, context in the header file is an incomplete type. A few other functions like the copy_to_gpu and copy_from_gpu are the same.

Comment (scxiao) [reply src/targets/gpu/include/migraphx/gpu/hip.hpp:155]: @pfultz2 the write_literals pass has an if branch like the following:
            if(enabled(MIGRAPHX_COPY_LITERALS{}))
            {
                literal l  = ins->get_literal();
                auto pre   = p.add_literal(l);
                auto alloc = p.insert_instruction(std::next(pre), hip_allocate{l.get_shape()});
                p.replace_instruction(ins, hip_copy_to_gpu{}, pre, alloc);
            }
            else

Should we also change this part?

Comment (pfultz2) [reply src/targets/gpu/include/migraphx/gpu/hip.hpp:155]: > Should we also change this part?

That shouldn't need to be changed.

**Pull request merged**